Microsoft.Toolkit.HighPerformance/Buffers/StringPool.cs

```diff
@@ -0,0 +1,199 @@
+        /// <param name="entriesPerBucket">The maximum number of <see cref="string"/> entries per bucket.</param>
+        public StringPool(int numberOfBuckets, int entriesPerBucket)
+        {
+            if (numberOfBuckets <= 0)
+            {
+                ThrowArgumentOutOfRangeException(nameof(numberOfBuckets));
+            }
+
+            if (entriesPerBucket <= 0)
+            {
+                ThrowArgumentOutOfRangeException(nameof(entriesPerBucket));
+            }
+
+            Span<Bucket> span = this.buckets = new Bucket[numberOfBuckets];
+
+            // We preallocate the buckets in advance, since each bucket only contains the
+            // array field, which is not preinitialized, so the allocations are minimal.
+            // This lets us lock on each individual buckets when retrieving a string instance.
+            foreach (ref Bucket bucket in span)
+            {
+                bucket = new Bucket();
+            }
+
+            NumberOfBuckets = numberOfBuckets;
+            EntriesPerBucket = entriesPerBucket;
+        }
+
+        /// <summary>
+        /// Gets the default <see cref="StringPool"/> instance.
+        /// </summary>
+        public static StringPool Default { get; } = new StringPool();
+
+        /// <summary>
+        /// Gets the total number of buckets in use.
+        /// </summary>
+        public int NumberOfBuckets { get; }
+
+        /// <summary>
+        /// Gets the maximum number of <see cref="string"/> entries stored in each bucket.
+        /// </summary>
+        public int EntriesPerBucket { get; }
+
+        /// <summary>
+        /// Gets a cached <see cref="string"/> instance matching the input content, or creates a new one.
+        /// </summary>
+        /// <param name="span">The input <see cref="ReadOnlySpan{T}"/> with the contents to use.</param>
+        /// <returns>A <see cref="string"/> instance with the contents of <paramref name="span"/>, cached if possible.</returns>
+        public string GetOrAdd(ReadOnlySpan<char> span)
+        {
+            if (span.IsEmpty)
+            {
+                return string.Empty;
+            }
+
+            int bucketIndex = span.Length % NumberOfBuckets;
+
+            Bucket bucket = this.buckets.DangerousGetReferenceAt(bucketIndex);
+
+            lock (bucket)
+            {
+                return bucket.GetOrAdd(span, EntriesPerBucket);
+            }
+        }
+
+        /// <summary>
+        /// Resets the current instance and its associated buckets.
+        /// </summary>
+        public void Reset()
+        {
+            foreach (Bucket bucket in this.buckets)
+            {
+                lock (bucket)
+                {
+                    bucket.Reset();
+                }
+            }
+        }
+
+        /// <summary>
+        /// A configurable bucket containing a group of cached <see cref="string"/> instances.
+        /// </summary>
+        private sealed class Bucket
```

Comment:
Haven't been thorough enough yet to work out if this would be an issue, but have you considered `struct`?

Reply:
I need these to be a class as I'm using `lock` over them eg. here:

https://github.com/windows-toolkit/WindowsCommunityToolkit/blob/77217a363476d656fbe734569f37f620f2008ff7/Microsoft.Toolkit.HighPerformance/Buffers/StringPool.cs#L106-L109

This is so that we can reduce the overhead with concurrent accesses and only lock per-bucket.

Reply:
I would suggest using a `struct` + a `SpinLock` (which lies and isn't actually a spinlock. it switches from spinning to yielding iirc), or a `struct` + `object` to lock on, to remove an unnecessary indirection and improve cache locality

Reply:
The [official docs for `SpinLock`](https://docs.microsoft.com/dotnet/api/system.threading.spinlock) explicitly suggest not to use them with potentially expensive operations, such as invoking code you don't own, making virtual calls or allocating memory - all of which we're doing inside the lock. I've switched the `Bucket` type to a `struct` in https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3380/commits/e2737e08e0cbe5d392a327ba63850c3cbbb22f00, let me know what you think of this new version! 😄